### PR TITLE
Add IWriter interfaces & dense Matrix/Vector implementations

### PR DESCRIPTION
**Pull request by DLehenbauer**
Nano gets IWriter, IVectorWriter, and IMatrixWriter interfaces (experimental)

Micro gets concrete implementations of DenseVector and RowMajorMatrix, which are writable producers backed by a dense JavaScript array.

### CODE REVIEW

File: packages/core/micro/src/consumerset.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
+export type ConsumerSet<T> = undefined | T | Array<{ value: T, count: number }>
```
Comment (DLehenbauer) [inline packages/core/micro/src/consumerset.ts:1]: Intended to provide efficient storage for a set of I*Consumers, but I'm dubious that it'll survive long.

I haven't benchmarked it against your much simpler `consumer0` solution, which is likely faster/simpler.

I'm also no longer convinced refcounting semantics are correct (mentioned in #58).

File: packages/core/micro/src/matrix/dense.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { MatrixProducer } from "./producer";
+
+/**
+ * A row/col-major agnostic base class 
+ */
+export abstract class DenseMatrix<T> extends MatrixProducer<T> {
```
Comment (DLehenbauer) [inline packages/core/micro/src/matrix/dense.ts:11]: (Another class that may not survive benchmarking.)

File: packages/core/micro/test/matrix.spec.ts [hunk 3]
```diff
@@ -0,0 +1,612 @@
...
+
+        matrix.insertRows(1, 1);
+        matrix.expect([
+            [0],
+            [undefined],
+            [1]
+        ]);
+    });
+
+    it("fail", () => {
```
Comment (DLehenbauer) [inline packages/core/micro/test/matrix.spec.ts:523]: Heh.  Forgot to give this test case a proper name.

File: packages/core/nano/src/types.ts [hunk 4]
```diff
@@ -198,6 +198,11 @@ export interface IReader<T> {
     readonly producer: IProducer<T>;
 }
 
+export interface IWriter<T> {
+    set<K extends keyof T>(property: K, value: T[K]): void;
+    delete<K extends keyof T>(property: K): void;
+}
```
Comment (DLehenbauer) [inline packages/core/nano/src/types.ts:204]: clear?

File: packages/core/nano/src/types.ts [hunk 5]
```diff
@@ -281,6 +291,10 @@ export interface IMatrixReader<T> {
     readonly matrixProducer: IMatrixProducer<T>;
 }
 
+export interface IMatrixWriter<T> {
+    setCell(row: number, col: number, value: T): void;
+}
```
Comment (DLehenbauer) [inline packages/core/nano/src/types.ts:296]: Note that an IMatrixWriter cannot insert/remove rows/cols.  I'm currently using IVectorWriters to change the matrix shape.

File: packages/core/micro/src/consumerset.ts (hunk 1 above)
Comment (jack-williams) [reply packages/core/micro/src/consumerset.ts:1]: > I haven't benchmarked it against your much simpler consumer0 solution, which is likely faster/simpler.

FWIW, I don't know if what I did was particularly good. I shameless ripped off https://github.com/janestreet/incremental

File: packages/core/nano/src/types.ts (hunk 5 above)
Comment (DLehenbauer) [reply packages/core/nano/src/types.ts:296]: Good idea.  :-)

File: packages/core/micro/src/consumerset.ts (hunk 1 above)
Comment (DLehenbauer) [reply packages/core/micro/src/consumerset.ts:1]: Interesting!

**Pull request merged**